src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,302 @@
+const MAX_RETRIES = 10;
+const MIN_BACKOFF = 100;
+const MAX_BACKOFF = 10 * 1000;
+const VERSION = 1;
+
+function sleep(ms: number) {
+	return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+function createTimestampFromDate(date = new Date()) {
+	const timestamp = new Timestamp();
+	timestamp.fromDate(date);
+	return timestamp;
+}
+
+function createTimestampFromMilliseconds(milliseconds: number) {
+	const MS_TO_NANOS = 1000000;
+	const seconds = Math.floor(milliseconds / 1000);
+	const nanos = (milliseconds - seconds * 1000) * MS_TO_NANOS;
+
+	const timestamp = new Timestamp();
+	timestamp.setSeconds(seconds);
+	timestamp.setNanos(nanos);
+	return timestamp;
+}
+
+export class LokiBackend implements DeviceLogsBackend {
+	private subscriptions: EventEmitter;
+	private querier: QuerierClient;
+	private pusher: PusherClient;
+	private ready: boolean;
+	private tailCalls: Map<string, grpc.ClientReadableStream<TailResponse>>;
+
+	constructor() {
+		this.subscriptions = new EventEmitter();
+		this.querier = new QuerierClient(
+			`${LOKI_HOST}:${LOKI_PORT}`,
+			createInsecureCredentials(),
+		);
+		this.pusher = new PusherClient(
+			`${LOKI_HOST}:${LOKI_PORT}`,
+			createInsecureCredentials(),
+		);
+		this.ready = false;
+		this.tailCalls = new Map();
+	}
+
+	public get available(): boolean {
+		return !this.ready;
+	}
+
+	/**
+	 *
+	 * Return $count of logs matching device_id in descending (BACKWARD) order.
+	 *
+	 * The logs are sorted by timestamp since Loki returns a distinct stream for each label combination.
+	 *
+	 * @param ctx
+	 * @param count
+	 */
+	public history(ctx: LogContext, count: number): Promise<DeviceLog[]> {
+		return new Promise((resolve, reject) => {
+			const oneHourAgo = new Date(Date.now() - 10000 * 60);
+
+			const queryRequest = new QueryRequest();
+			queryRequest.setSelector(this.getDeviceQuery(ctx));
+			queryRequest.setLimit(Number.isFinite(count) ? count : 1000);
+			queryRequest.setStart(createTimestampFromDate(oneHourAgo));
+			queryRequest.setEnd(createTimestampFromDate());
+			queryRequest.setDirection(Direction.BACKWARD);
+
+			const deviceLogs: DeviceLog[] = [];
+			const call = this.querier.query(
+				queryRequest,
+				createOrgIdMetadata(String(ctx.belongs_to__application)),
+			);
+			call.on('data', (queryResponse: QueryResponse) => {
+				deviceLogs.push(
+					...this.fromStreamsToDeviceLogs(queryResponse.getStreamsList()),
+				);
+			});
+			call.on('error', (error: Error & { details: string }) => {
+				reject(new BadRequestError(`Query failed: ${error.details}"`));
+			});
+			call.on('end', () => {
+				resolve(_.orderBy(deviceLogs, 'timestamp', 'desc'));
+			});
+		});
+	}
+
+	public async publish(
+		ctx: LogWriteContext,
+		logs: Array<DeviceLog & { version?: number }>,
+	): Promise<any> {
+		const streams = this.fromDeviceLogsToStreams(ctx, logs);
+		try {
+			let tries = 0;
+			let nextBackoff = MIN_BACKOFF;
+			let prevBackoff = MIN_BACKOFF;
+			while (nextBackoff <= MAX_BACKOFF) {
```

Comment:
Why do we have both a max retries and max backoff? I would expect only one can ever apply in reality and so doing
```suggestion
			for (let tries = 0; tries < MAX_BACKOFF; tries++) {
```
would be more straightforward, and then you can also use
```ts
await Bluebird.delay(Math.min(MAX_BACKOFF, nextBackoff));
```
to cap the backoff to reasonable levels

Reply:
Yep `tries` and `MAX_BACKOFF` was overly redundant. Removed the `tries` logic but kept the back-off logic the same. Let me know if its acceptable.